### PR TITLE
feat: new built-in aggregate function median

### LINKED ISSUE #1960: feat: new built-in aggregate function `median`
You may follow the built-in aggregate function development doc (Chinese)

https://openmldb.ai/docs/zh/main/developer/built_in_function_develop_guide.html

### PR DESCRIPTION
**Pull request by wuxiaobai24**
* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
Add a new build-in aggregate function `median`.

* **What is the current behavior?** (You can also link to an open issue here)
It should resolve #1960 

* **What is the new behavior (if this is a feature change)?**
a new build-in aggregate function `median`.

### CODE REVIEW

File: hybridse/src/udf/default_udf_library.cc [hunk 1]
```diff
@@ -310,6 +310,47 @@ struct DistinctCountDef {
...
+            std::sort(vec->begin(), vec->end());
+            size_t mid = vec->size() / 2;
+            if (vec->size() % 2 == 0) {
+                *ret = (vec->at(mid - 1) + vec->at(mid)) / 2.0;
+            } else {
+                *ret = vec->at(mid);
+            }
+        }
+        vec->clear();
+        vec->~VectorT();
```
Comment (aceforeverd) [inline hybridse/src/udf/default_udf_library.cc]: I guess it is ok to just `delete` ?

File: hybridse/src/udf/default_udf_library.cc [hunk 2]
```diff
@@ -310,6 +310,47 @@ struct DistinctCountDef {
...
+        return vec;
+    }
+
+    static void Output(VectorT* vec, double* ret, bool* is_null) {
+        if (vec->empty()) {
+            *is_null = true;
+        } else {
+            *is_null = false;
+
+            std::sort(vec->begin(), vec->end());
```
Comment (aceforeverd) [inline hybridse/src/udf/default_udf_library.cc]: using two heap structure in update help avoid sort

File: hybridse/src/udf/default_udf_library.cc [hunk 3]
```diff
@@ -310,6 +310,47 @@ struct DistinctCountDef {
     };
 };
 
+template <typename T>
+struct MedianDef {
+    using ArgT = typename DataTypeTrait<T>::CCallArgType;
+    using VectorT = std::vector<T>;
+
+    void operator()(UdafRegistryHelper& helper) {
```
Comment (aceforeverd) [inline hybridse/src/udf/default_udf_library.cc]: add `// NOLINT` at end

File: hybridse/src/udf/default_udf_library.cc [hunk 4]
```diff
@@ -2434,6 +2475,29 @@ void DefaultUdfLibrary::InitUdaf() {
...
+            |--|
+            |1|
+            |2|
+            |3|
+            |4|
+            @code{.sql}
+                SELECT median(value) OVER w;
+                -- output 2.5
+            @endcode
+            @since 0.5.0
```
Comment (aceforeverd) [inline hybridse/src/udf/default_udf_library.cc]: since 0.6.0

File: hybridse/src/udf/default_udf_library.cc [hunk 5]
```diff
@@ -310,6 +310,47 @@ struct DistinctCountDef {
     };
 };
 
+template <typename T>
+struct MedianDef {
```
Comment (aceforeverd) [inline hybridse/src/udf/default_udf_library.cc:316]: `median` needs end-to-end verify. 
May you try add a new case in `cases/function/function/test_udaf_function.yaml`, you can see the example like https://github.com/4paradigm/OpenMLDB/blob/af0272cec5b9646c8b7eaa3c6d9e9dfccaf6f56d/cases/function/function/test_udaf_function.yaml#L114-L143.  Docs is missing so don't afraid to ask.

Finally you can build the target `toydb_run_engine` and verify your newly case via:
```sh
./build/hybridse/examples/toydb/src/toydb_run_engine --yaml_path cases/function/function/test_udaf_function.yaml --case_id 4
```

I would be nice if you can add it

File: hybridse/src/udf/default_udf_library.cc (hunk 1 above)
Comment (wuxiaobai24) [reply hybridse/src/udf/default_udf_library.cc]: > I guess it is ok to just `delete` ?

This is what I did in the first place, but it leads to double free in testing. So I referred to the code in `DistinctCountDef` and `TopKDef`

https://github.com/4paradigm/OpenMLDB/blob/af0272cec5b9646c8b7eaa3c6d9e9dfccaf6f56d/hybridse/src/udf/default_udf_library.cc#L292

https://github.com/4paradigm/OpenMLDB/blob/af0272cec5b9646c8b7eaa3c6d9e9dfccaf6f56d/hybridse/src/udf/containers.h#L87

File: hybridse/src/udf/default_udf_library.cc (hunk 2 above)
Comment (wuxiaobai24) [reply hybridse/src/udf/default_udf_library.cc]: > using two heap structure in update help avoid sort

Wow, good idea.

File: hybridse/src/udf/default_udf_library.cc (hunk 3 above)
Comment (wuxiaobai24) [reply hybridse/src/udf/default_udf_library.cc]: > add `// NOLINT` at end

OK!

File: hybridse/src/udf/default_udf_library.cc (hunk 4 above)
Comment (wuxiaobai24) [reply hybridse/src/udf/default_udf_library.cc]: > since 0.6.0

OK!

File: hybridse/src/udf/default_udf_library.cc (hunk 5 above)
Comment (wuxiaobai24) [reply hybridse/src/udf/default_udf_library.cc:316]: > `median` needs end-to-end verify. May you try add a new case in `cases/function/function/test_udaf_function.yaml`, you can see the example like
> 
> https://github.com/4paradigm/OpenMLDB/blob/af0272cec5b9646c8b7eaa3c6d9e9dfccaf6f56d/cases/function/function/test_udaf_function.yaml#L114-L143
> 
> . Docs is missing so don't afraid to ask.
> Finally you can build the target `toydb_run_engine` and verify your newly case via:
> 
> ```shell
> ./build/hybridse/examples/toydb/src/toydb_run_engine --yaml_path cases/function/function/test_udaf_function.yaml --case_id 4
> ```
> 
> I would be nice if you can add it
ok, i try to add a end-to-end test case.

**Review by aceforeverd — APPROVED** [review]

File: hybridse/src/udf/default_udf_library.cc [hunk 6]
```diff
@@ -310,6 +311,71 @@ struct DistinctCountDef {
...
+    using ArgT = typename DataTypeTrait<T>::CCallArgType;
+    using MaxHeapT = std::priority_queue<T, std::vector<T>, std::less<>>;
```
Comment (zhanghaohit) [inline hybridse/src/udf/default_udf_library.cc]: cpp lint error. kindly solve all the lint warnings.

**Review by zhanghaohit — APPROVED** [review]
LGTM. Thanks @wuxiaobai24 for the contribution!

**Pull request merged**